ding/data/shm_buffer.py

```diff
@@ -18,8 +21,82 @@
     np.float64: ctypes.c_double,
 }
 
+# uint16, uint32, uint32
+_NTYPE_TO_TTYPE = {
+    np.bool_: torch.bool,
+    np.uint8: torch.uint8,
+    # np.uint16: torch.int16,
+    # np.uint32: torch.int32,
+    # np.uint64: torch.int64,
+    np.int8: torch.uint8,
+    np.int16: torch.int16,
+    np.int32: torch.int32,
+    np.int64: torch.int64,
+    np.float32: torch.float32,
+    np.float64: torch.float64,
+}
+
+_NOT_SUPPORT_NTYPE = {np.uint16: torch.int16, np.uint32: torch.int32, np.uint64: torch.int64}
+_CONVERSION_TYPE = {np.uint16: np.int16, np.uint32: np.int32, np.uint64: np.int64}
+
+
+class ShmBufferBase:
+
+    def fill(self, src_arr: np.ndarray) -> None:
+        raise NotImplementedError
+
+    def get(self) -> np.ndarray:
+        raise NotImplementedError
+
 
-class ShmBuffer():
+class ShmBufferCuda(ShmBufferBase):
+
+    def __init__(
+            self,
+            dtype: Union[type, np.dtype],
+            shape: Tuple[int],
+            copy_on_get: bool = True,
+            ctype: Optional[type] = None,
+            device: Optional[torch.device] = torch.device('cuda:0')
+    ) -> None:
+        if isinstance(dtype, np.dtype):  # it is type of gym.spaces.dtype
+            dtype = dtype.type
+        if dtype in _NOT_SUPPORT_NTYPE.keys():
+            logging.warning(
+                "Torch tensor unsupport numpy type {}, attempt to do a type conversion, which may lose precision.".
+                format(dtype)
+            )
+            ttype = _NOT_SUPPORT_NTYPE[dtype]
+            self.dtype = _CONVERSION_TYPE[dtype]
+        else:
+            ttype = _NTYPE_TO_TTYPE[dtype]
+            self.dtype = dtype
+        self.copy_on_get = copy_on_get
+        self.shape = shape
+        self.device = device
+        self.buffer = torch.zeros(reduce(lambda x, y: x * y, shape), dtype=ttype, device=self.device)
+
+    def fill(self, src_arr: np.ndarray) -> None:
+        assert isinstance(src_arr, np.ndarray), type(src_arr)
+
+        if src_arr.dtype.type != self.dtype:
+            logging.warning(
+                "Torch tensor unsupport numpy type {}, attempt to do a type conversion, which may lose precision.".
+                format(self.dtype)
+            )
+            src_arr = src_arr.astype(self.dtype)
+
+        cpu_t = torch.from_numpy(src_arr)
+        self.buffer.copy_(cpu_t.view(cpu_t.numel()), False)
```

Comment:
why `non_blocking=False` here

Reply:
if `non_blocking ` is `True `and this copy_ is between CPU and GPU, the copy_ may occur asynchronously with respect to the host. 
`False ` is the default value of `non_blocking `, should we change its' value to `True` for better performance?